lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -604,6 +604,23 @@ findMemberReference(Expr *expr) {
   return None;
 }
 
+/// Return true if the callee of an ApplyExpr is async
+///
+/// Note that this must be called after the implicitlyAsync flag has been set,
+/// or implicitly async calls will not return the correct value.
+static bool isAsyncCall(const ApplyExpr *call) {
+  if (call->implicitlyAsync())
+    return true;
+  // Effectively the same as doing a
+  // `dyn_cast_or_null<FunctionType>(call->getFn()->getType())`, check the
+  // result of that and then checking `isAsync` if it's defined.
+  Type funcTypeType = call->getFn()->getType();
+  if (!funcTypeType)
+    return false;
+  FunctionType *funcType = funcTypeType->getAs<FunctionType>();
```

Comment:
I think this could be converted into a `castTo<FunctionType>()` because otherwise it would never be a correct argument application (I'm assuming that this is checked after expression has been successfully type-checked).

Reply:
Changed